docs/WIA.md

```diff
@@ -0,0 +1,234 @@
+ - For the compression methods NONE and PURGE, if the end offset of the last ``wia_except_list_t`` is not evenly divisible by 4, padding is inserted after it so that the data afterwards will start at a 4 byte boundary. This padding is not inserted for the other compression methods.
+
+## `wia_segment_t`
+
+This struct is used by the simple compression method PURGE, which stores runs of zeroes efficiently and stores other data as is.
+
+|Type and name|Description|
+|--|--|
+|`u32 offset`|The offset of `data` within the decompressed data. (Any `wia_except_list_t` structs are not counted as part of the decompressed data.)|
+|`u32 size`|The number of bytes in `data`.|
+|`u8 data[size]`|Data.|
+
+Each PURGE chunk contains one or more `wia_segment_t` structs stored in order of ascending `offset`, followed by a SHA-1 hash (0x14 bytes) of the `wia_except_list_t` structs (if any) and the `wia_segment_t` structs. Bytes in the decompressed data that are not covered by any `wia_segment_t` struct are set to `0x00`.
+
+# RVZ file format description
+
+RVZ is a file format which is closely based on WIA. The differences are as follows:
+
+* Zstandard has been added as a compression method. `compression` in `wia_disc_t` is set to 5 when Zstandard is used, and there is no compressor specific data. `compr_level` in `wia_disc_t` should be treated as signed instead of unsigned because Zstandard supports negative compression levels.
+* PURGE has been removed as a compression method.
+* Chunk sizes smaller than 2 MiB are supported. The following applies when using a chunk size smaller than 2 MiB:
+    * The chunk size must be at least 32 KiB and must be a power of two. (Just like with WIA, sizes larger than 2 MiB do not have to be a power of two, they just have to be an integer multiple of 2 MiB.)
+    * For Wii partition data, each chunk contains one `wia_except_list_t` which contains exceptions for that chunk (and no other chunks). Offset 0 refers to the first hash of the current chunk, not the first hash of the full 2 MiB of data.
+* An encoding scheme which is described below is used to store pseudorandom padding data losslessly.
+
+## RVZ packing
+
+The RVZ packing encoding scheme is applied to all `wia_group_t` data, with any bzip2/LZMA/Zstandard compression being applied on top of it. (In other words, when reading an RVZ file, bzip2/LZMA/Zstandard decompression is done before decoding the RVZ packing.) RVZ packed data can be decoded as follows:
+
+1. Read 4 bytes of data and interpret it as a 32-bit unsigned big endian integer. Call this `size`.
+2. If the most significant bit of `size` is not set, read `size` bytes and output them unchanged. If the most significant bit of `size` is set, unset the most significant bit of `size`, then read 68 bytes of PRNG seed data and output `size` bytes using the PRNG algorithm described below.
+3. Repeat until all input has been read.
```

Comment:
I also miss some kind of decompression header. AFAIK crompression libraries assumes that the caller checks if the size actually goes down. For 32 kB chunks, ZStd has an upper bound of 176 additional bytes. 512 bytes for 128 kB chunks. How often is `compressed_size > raw_size`?
Maybe we can create an RVZ header here with flags like: compressed, contains_PRNG_packing, with_except_list. Or is there a way to extend the wia_group_t struct?

Reply:
> Is there a field for the size of the compressed but not unpacked data?

No.

> Else I assume you rely on the outcome of the compression library. iirc some compression methods do not store the size of the uncompressed data, so this might make it tricky to extend the RVZ format afterwards.

This is exactly correct.

> It would be nice to store the uncompressed-but-unpacked size somewhere as well.

I'll give this a try sometime (extending `wia_group_t` like you mentioned on IRC). I'll have to see if it makes the code messier...

Reply:
> I also miss some kind of decompression header.

Do you mean a flag for whether a chunk is compressed or not?

> ZStd has an upper bound of 176 additional bytes. 512 bytes for 128 kB chunks.

Where do those numbers come from? Based on reading RFC 8478 a little (I haven't read all of it), it seems like the overhead would be smaller than that, at least judging by this part:

```
   A Compressed_Block has the extra restriction that Block_Size is
   always strictly less than the decompressed size.  If this condition
   cannot be respected, the block must be sent uncompressed instead
   (i.e., treated as a Raw_Block).
```

> How often is `compressed_size > raw_size`?

I'd imagine it would be quite common in the case where the game already has applied its own compression to the data.

Reply:
ZSTD_COMPRESSBOUND in https://facebook.github.io/zstd/zstd_manual.html

Your cited block is not possible. No method can compress arbitary data. In fact, even on the average case, they can not result in a smaller file size. The amount of possible states of the arbitary data must still fit in the output format. So they usually assume that the data has some kind of pattern, and ask you to not use the method else. However this needs to be encoded somewhere. The information "the block must be sent uncompressed instead" must be encoded somewhere, and this is the additional flag I'm asking for.

Reply:
Yes, I am not claiming that Zstandard never has any overhead, I'm only claiming that the overhead is smaller than what you were claiming (the size of the Zstandard frame header plus 4 bytes per Zstandard block). So there is still a point to your flag that indicates whether the data is compressed.